File: src/main/java/no/novari/flyt/configuration/model/configuration/dtos/ValueMappingDto.java
package no.novari.flyt.configuration.model.configuration.dtos;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.extern.jackson.Jacksonized;
import no.novari.flyt.configuration.model.configuration.entities.ValueMapping;
import no.novari.flyt.configuration.validation.constraints.InstanceValueKeysAreDefinedInMetadata;
import no.novari.flyt.configuration.validation.constraints.InstanceValueTypesAreCompatible;
import no.novari.flyt.configuration.validation.constraints.ValueParsableAsType;
import no.novari.flyt.configuration.validation.groups.InstanceValueKeys;
import no.novari.flyt.configuration.validation.groups.InstanceValueTypes;
import no.novari.flyt.configuration.validation.groups.ValueParsability;

import jakarta.validation.constraints.NotNull;

@ValueParsableAsType(groups = ValueParsability.class)
@InstanceValueKeysAreDefinedInMetadata(groups = InstanceValueKeys.class)
@InstanceValueTypesAreCompatible(groups = InstanceValueTypes.class)
@Getter
@Builder
@EqualsAndHashCode
@Jacksonized
public class ValueMappingDto {

    @NotNull
    private final ValueMapping.Type type;

    @NotNull
    private final String mappingString;

}

File: src/main/kotlin/no/novari/flyt/configuration/model/configuration/dtos/ValueMappingDto.kt
package no.novari.flyt.configuration.model.configuration.dtos

import jakarta.validation.constraints.NotNull
import no.novari.flyt.configuration.model.configuration.entities.ValueMapping
import no.novari.flyt.configuration.validation.constraints.InstanceValueKeysAreDefinedInMetadata
import no.novari.flyt.configuration.validation.constraints.InstanceValueTypesAreCompatible
import no.novari.flyt.configuration.validation.constraints.ValueParsableAsType
import no.novari.flyt.configuration.validation.groups.InstanceValueKeys
import no.novari.flyt.configuration.validation.groups.InstanceValueTypes
import no.novari.flyt.configuration.validation.groups.ValueParsability

@ValueParsableAsType(groups = [ValueParsability::class])
@InstanceValueKeysAreDefinedInMetadata(groups = [InstanceValueKeys::class])
@InstanceValueTypesAreCompatible(groups = [InstanceValueTypes::class])
data class ValueMappingDto(
    @field:NotNull
    var type: ValueMapping.Type? = null,
    @field:NotNull
    var mappingString: String? = null,
) {
    companion object {
        @JvmStatic
        fun builder(): Builder = Builder()
    }

    class Builder {
        private var type: ValueMapping.Type? = null
        private var mappingString: String? = null

        fun type(type: ValueMapping.Type?) = apply { this.type = type }

        fun mappingString(mappingString: String?) = apply { this.mappingString = mappingString }

        fun build(): ValueMappingDto = ValueMappingDto(type = type, mappingString = mappingString)
    }
}
